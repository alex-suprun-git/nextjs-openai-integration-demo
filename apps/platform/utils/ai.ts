import { getLocale } from 'next-intl/server';
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import { OutputFixingParser, StructuredOutputParser } from 'langchain/output_parsers';
import { PromptTemplate } from '@langchain/core/prompts';
import { Document } from 'langchain/document';
import { loadQARefineChain } from 'langchain/chains';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import z from 'zod';

const modelParams = {
  model: 'o4-mini',
};

const parser = StructuredOutputParser.fromZodSchema(
  z.object({
    mood: z.string().describe('the mood of the person who wrote the journal entry.'),
    subject: z.string().describe('the subject of the journal entry.'),
    negative: z
      .boolean()
      .describe('is the journal entry negative? (i.e. does it contain negative emotions?).'),
    summary: z.string().max(100).describe('quick summary of the entire entry.'),
    color: z
      .string()
      .describe(
        'a hexidecimal color code that represents the mood of the entry. Example #0101fe for blue representing happiness.',
      ),
    sentimentScore: z
      .number()
      .describe(
        'sentiment of the text and rated on a scale from -10 to 10, where -10 is extremely negative, 0 is neutral, and 10 is extremely positive.',
      ),
    title: z.string().describe('A concise title summarizing the journal entry.'),
  }) as any,
);

const getPrompt = async (content: string, language: UserLocale) => {
  const formatInstructions = parser.getFormatInstructions();

  const templates = {
    de: 'Analysiere den folgenden Tagebucheintrag und antworte auf Deutsch. Übersetze die Formatierungsanweisungen ins Deutsche und befolge sie unbedingt! \n{formatInstructions}\n{entry}',
    en: 'Analyze the following journal entry and respond in English. Follow the instructions and format your response to match the format instructions, no matter what! \n{formatInstructions}\n{entry}',
  };

  const prompt = new PromptTemplate({
    template: templates[language],
    inputVariables: ['entry'],
    partialVariables: { formatInstructions },
  });

  const input = await prompt.format({
    entry: content,
  });

  return input;
};

export const analyzeEntry = async (content: string) => {
  const locale = await getLocale();

  const input = await getPrompt(content, locale as UserLocale);
  const model = new ChatOpenAI({
    ...modelParams,
    apiKey: process.env.OPENAI_API_KEY,
  });
  const output = await model.invoke(input);

  try {
    return await parser.parse(output.content as string);
  } catch (e) {
    const fixParser = OutputFixingParser.fromLLM(new ChatOpenAI(modelParams), parser);
    const fix = await fixParser.parse(output.content as string);
    return fix;
  }
};

export const analysisFeedback = async (
  language: UserLocale,
  question: string,
  entries: BaseEntry[],
) => {
  if (!entries.length) {
    return undefined;
  }

  const docs = entries.map(
    (entry) =>
      new Document({
        pageContent: entry.content as string,
        metadata: { source: entry.id, date: entry.createdAt },
      }),
  );
  const model = new ChatOpenAI(modelParams);
  const chain = loadQARefineChain(model);
  const embeddings = new OpenAIEmbeddings();
  const store = await MemoryVectorStore.fromDocuments(docs, embeddings);
  const relevantDocs = await store.similaritySearch(question);

  const templates = {
    en: `Answer the following question focusing exclusively on the person's mood and feelings based on the given documents. If the question touches on any other topic or if the context is unclear, please respond with a polite request to rephrase the question. Ensure your answer is in the same language as the input. \nQuestion: {question}`,
    de: `Beantworte die folgende Frage ausschließlich im Hinblick auf die Stimmung und Gefühle der Person basierend auf den gegebenen Dokumenten. Falls die Frage ein anderes Thema betrifft oder der Kontext unklar ist, bitte höflich um eine Umformulierung der Frage. Stelle sicher, dass deine Antwort in derselben Sprache wie die Eingabe erfolgt. \nFrage: {question}`,
  };

  const promptTemplate = new PromptTemplate({
    template: templates[language],
    inputVariables: ['question'],
  });

  const input = await promptTemplate.format({
    question: question,
  });

  const res = await chain.invoke({
    input_documents: relevantDocs,
    question: input,
  });

  return res.output_text;
};
