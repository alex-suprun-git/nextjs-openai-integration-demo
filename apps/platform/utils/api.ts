export const createUrl = (path: string) => window.location.origin + path;

export const updateUserPromptUsage = async (promptContentLength: number) => {
  const res = await fetch(
    new Request('/api/user', {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ promptContentLength }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const createNewEntry = async (content: string) => {
  const res = await fetch(
    new Request('/api/journal', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

// @Deprecated
export const updateEntry = async (id: string, content: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const deleteEntry = async (id: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  return;
};

export const askQuestion = async (question: string) => {
  const res = await fetch(
    new Request('/api/question', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ question }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data;
};
