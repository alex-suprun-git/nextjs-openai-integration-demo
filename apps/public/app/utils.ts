export const createUrl = (path: string) => window.location.origin + path;

export const getPageURL = (url: string) => {
  const [, , ...restSegments] = url.split('/');
  return restSegments.join('/');
};

export async function fetcher(_url: string) {
  const url = createUrl(_url);

  const res = await fetch(
    new Request(url, {
      method: 'GET',
    }),
  );

  if (!res.ok) return;
  return res.json() as Promise<{ isSignedIn: boolean }>;
}
