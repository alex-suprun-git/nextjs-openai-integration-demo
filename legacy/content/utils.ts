import { contentGQLQuery } from './fetch';

export const getContentFromCMS = async (
  query: string,
  locale: string,
  extraVariables: Record<string, any> = {},
) => {
  const variables = { locale, ...extraVariables };

  try {
    const data = await contentGQLQuery({ query, variables });

    if (!data) {
      throw new Error('Error fetching hero content');
    }

    return data;
  } catch (error) {
    console.error(error);
    return null;
  }
};
