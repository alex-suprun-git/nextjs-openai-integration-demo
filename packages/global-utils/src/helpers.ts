import { Environment } from './types';

export function getCurrentEnv(options: { host?: string } = {}): Environment {
	if (typeof window !== 'undefined') {
		const host = window.location.hostname;
		if (host === 'localhost' || host.endsWith('.local')) {
			return Environment.DEV;
		}
		return Environment.PROD;
	}

	if (options.host) {
		const host = options.host.split(':')[0];
		if (host === 'localhost' || host === '127.0.0.1') {
			return Environment.DEV;
		}
		return Environment.PROD;
	}

	return process.env.NODE_ENV === Environment.PROD
		? Environment.PROD
		: Environment.DEV;
}
