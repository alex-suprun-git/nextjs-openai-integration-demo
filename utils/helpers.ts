export const getExcerpt = (content: string): string =>
  content.length >= 100 ? `${content.slice(0, 100)}...` : content;

export const formatDate = (date: Date): string => {
  const userTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone;

  return new Intl.DateTimeFormat('en-GB', {
    dateStyle: 'medium',
    timeStyle: 'short',
    timeZone: userTimezone,
  }).format(date);
};

export const convertHexToRGBA = (_hex: string, _opacity: number = 1) => {
  let opacity = _opacity;
  let hex = _hex.replace('#', '');

  if (hex.length === 3) {
    hex = `${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}`;
  }

  const r = parseInt(hex.substring(0, 2), 16);
  const g = parseInt(hex.substring(2, 4), 16);
  const b = parseInt(hex.substring(4, 6), 16);

  /* Backward compatibility for whole number based opacity values. */
  if (opacity > 1 && opacity <= 100) {
    opacity = opacity / 100;
  } else if (opacity > 100) opacity = 1;

  return `rgba(${r},${g},${b},${opacity})`;
};
