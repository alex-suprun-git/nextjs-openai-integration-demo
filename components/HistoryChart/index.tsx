'use client';
import { formatDate } from '@/utils/helpers';
import { ReactElement } from 'react';
import { ResponsiveContainer, LineChart, Line, XAxis, Tooltip, TooltipProps } from 'recharts';

const CustomTooltip = ({
  active,
  payload,
  label,
}: TooltipProps<number, string>): ReactElement | null => {
  if (active && payload?.length) {
    const analysis = payload[0].payload;

    return (
      <div className="custom-tooltip relative rounded-lg border border-black/10 bg-white/5 p-8 shadow-md backdrop-blur-md">
        <div
          className="absolute left-2 top-2 h-2 w-2 rounded-full"
          style={{ background: analysis.color }}
        ></div>
        <p className="label text-sm">{label}</p>
        <p className="intro text-xl uppercase">{analysis.mood}</p>
      </div>
    );
  }

  return null;
};

const HistoryChart = ({ data }: { data: AnalysisEntry[] }) => {
  const formattedData = data.map((entry) => ({
    ...entry,
    updatedAt: formatDate(new Date(entry.updatedAt)),
  }));

  return (
    <div className="h-full w-full">
      <ResponsiveContainer aspect={3}>
        <LineChart width={300} height={100} data={formattedData}>
          <Line
            type="monotone"
            dataKey="sentimentScore"
            stroke="#8884d8"
            strokeWidth={2}
            activeDot={{ r: 8 }}
          />
          <XAxis dataKey="updatedAt" />
          <Tooltip content={<CustomTooltip />} />
        </LineChart>
      </ResponsiveContainer>
    </div>
  );
};

export default HistoryChart;
