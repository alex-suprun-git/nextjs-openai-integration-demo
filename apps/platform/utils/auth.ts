import { getServerSession } from 'next-auth';
import { authOptions } from '@/app/api/auth/[...nextauth]/auth';
import { prisma } from './db';
import { Prisma } from '@prisma/client';
import { isDynamicServerError } from 'next/dist/client/components/hooks-server-context';

export const getCurrentUser = async () => {
  try {
    const session = await getServerSession(authOptions);

    if (!session?.user?.email) {
      console.warn('User email is not available. User might be signed out.');
      return null;
    }

    const user = await prisma.user.findFirstOrThrow({
      where: {
        email: session.user.email as string,
      },
    });

    return user;
  } catch (error: any) {
    if (error instanceof Prisma.PrismaClientValidationError) {
      console.error('Prisma validation error:', error.message);
      throw new Error('Failed to find user due to validation error.');
    } else if (error.code === 'P2025') {
      console.error('User not found:', error.message);
      throw new Error('User not found.');
    } else {
      if (isDynamicServerError(error)) {
        throw error;
      }
      console.error('Unexpected error:', error);
      throw new Error('An unexpected error occurred.');
    }
  }
};
