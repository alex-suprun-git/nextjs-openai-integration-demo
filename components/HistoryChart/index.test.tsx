import { render, screen, within } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import HistoryChart from '@/components/HistoryChart';
import { formatDate } from '@/utils/helpers';

class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

vi.mock('recharts', async (importOriginal) => {
  const originalModule = (await importOriginal()) as Record<string, unknown>;
  return {
    ...originalModule,
    ResponsiveContainer: ({ children }: any) => <div>{children}</div>,
  };
});

beforeEach(() => {
  (global as any).ResizeObserver = ResizeObserver;
});

describe('HistoryChart', () => {
  const mockData = [
    { sentimentScore: 1, updatedAt: '2023-01-01T00:00:00Z', mood: 'Happy', color: '#ff0000' },
    { sentimentScore: 0, updatedAt: '2023-01-02T00:00:00Z', mood: 'Neutral', color: '#00ff00' },
  ] as any[];

  it('renders correctly and matches snapshot', () => {
    const { container } = render(<HistoryChart data={mockData} />);
    expect(container).toMatchSnapshot();
  });

  it('renders tooltip correctly when active and data is present', async () => {
    render(<HistoryChart data={mockData} />);

    // Simulate the Tooltip being active
    const activeTooltipData = {
      sentimentScore: 1,
      updatedAt: '2023-01-01',
      mood: 'Happy',
      color: '#ff0000',
    };

    render(
      <div
        data-testid="custom-tooltip"
        className="custom-tooltip relative rounded-lg border border-black/10 bg-white/5 p-8 shadow-md backdrop-blur-md"
      >
        <div
          className="absolute left-2 top-2 h-2 w-2 rounded-full"
          style={{ background: activeTooltipData.color }}
        ></div>
        <p className="label text-sm">{activeTooltipData.updatedAt}</p>
        <p className="intro text-xl uppercase">{activeTooltipData.mood}</p>
      </div>,
    );

    const tooltipElement = screen.getByTestId('custom-tooltip');
    expect(tooltipElement).toBeInTheDocument();

    const labelElement = within(tooltipElement!).getByText('2023-01-01');
    expect(labelElement).toBeInTheDocument();

    const moodElement = within(tooltipElement!).getByText('Happy');
    expect(moodElement).toBeInTheDocument();
  });

  it('formats data correctly before rendering', () => {
    render(<HistoryChart data={mockData} />);

    mockData.forEach((entry) => {
      const formattedDate = formatDate(new Date(entry.updatedAt));
      expect(screen.getByText(formattedDate)).toBeInTheDocument();
    });
  });

  it('renders correctly with empty data', () => {
    const { container } = render(<HistoryChart data={[]} />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByText(/2023-01-01/i)).not.toBeInTheDocument();
  });
});
