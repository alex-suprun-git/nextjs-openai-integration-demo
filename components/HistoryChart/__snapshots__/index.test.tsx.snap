// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HistoryChart > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="h-full w-full"
  >
    <div>
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 300px; height: 100px;"
      >
        <svg
          class="recharts-surface"
          height="100"
          style="width: 100%; height: 100%;"
          viewBox="0 0 300 100"
          width="300"
        >
          <title />
          <desc />
          <defs>
            <clippath
              id="recharts1-clip"
            >
              <rect
                height="60"
                width="290"
                x="5"
                y="5"
              />
            </clippath>
          </defs>
          <g
            class="recharts-layer recharts-line"
          >
            <path
              class="recharts-curve recharts-line-curve"
              d="M5,5L295,65"
              fill="none"
              height="60"
              stroke="#8884d8"
              stroke-dasharray="0px 0px"
              stroke-width="2"
              width="290"
            />
          </g>
          <g
            class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
          >
            <line
              class="recharts-cartesian-axis-line"
              fill="none"
              height="30"
              orientation="bottom"
              stroke="#666"
              width="290"
              x="5"
              x1="5"
              x2="295"
              y="65"
              y1="65"
              y2="65"
            />
            <g
              class="recharts-cartesian-axis-ticks"
            >
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="30"
                  orientation="bottom"
                  stroke="#666"
                  width="290"
                  x="5"
                  x1="5"
                  x2="5"
                  y="65"
                  y1="71"
                  y2="65"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="30"
                  orientation="bottom"
                  stroke="none"
                  text-anchor="middle"
                  width="290"
                  x="5"
                  y="73"
                >
                  <tspan
                    dy="0.71em"
                    x="5"
                  >
                    1 Jan 2023, 01:00
                  </tspan>
                </text>
              </g>
              <g
                class="recharts-layer recharts-cartesian-axis-tick"
              >
                <line
                  class="recharts-cartesian-axis-tick-line"
                  fill="none"
                  height="30"
                  orientation="bottom"
                  stroke="#666"
                  width="290"
                  x="5"
                  x1="295"
                  x2="295"
                  y="65"
                  y1="71"
                  y2="65"
                />
                <text
                  class="recharts-text recharts-cartesian-axis-tick-value"
                  fill="#666"
                  height="30"
                  orientation="bottom"
                  stroke="none"
                  text-anchor="middle"
                  width="290"
                  x="295"
                  y="73"
                >
                  <tspan
                    dy="0.71em"
                    x="295"
                  >
                    2 Jan 2023, 01:00
                  </tspan>
                </text>
              </g>
            </g>
          </g>
        </svg>
        <div
          class="recharts-tooltip-wrapper"
          style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HistoryChart > renders correctly with empty data 1`] = `
<div>
  <div
    class="h-full w-full"
  >
    <div>
      <div
        class="recharts-wrapper"
        style="position: relative; cursor: default; width: 300px; height: 100px;"
      >
        <svg
          class="recharts-surface"
          height="100"
          style="width: 100%; height: 100%;"
          viewBox="0 0 300 100"
          width="300"
        >
          <title />
          <desc />
          <defs>
            <clippath
              id="recharts10-clip"
            >
              <rect
                height="60"
                width="290"
                x="5"
                y="5"
              />
            </clippath>
          </defs>
        </svg>
        <div
          class="recharts-tooltip-wrapper"
          style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</div>
`;
